airflow-core/src/airflow/ui/src/hooks/useGridNavigation.ts

```diff
@@ -0,0 +1,129 @@
+/*!
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { useCallback, useEffect } from "react";
+import { useNavigate, useParams } from "react-router-dom";
+
+import type { GridTask, RunWithDuration } from "src/layouts/Details/Grid/utils";
+
+type UseGridNavigationProps = {
+  flatNodes: Array<GridTask>;
+  isGridFocused: boolean;
+  runs: Array<RunWithDuration>;
+};
+
+export const useGridNavigation = ({ flatNodes, isGridFocused, runs }: UseGridNavigationProps) => {
+  const navigate = useNavigate();
+  const { dagId = "", runId = "", taskId = "" } = useParams();
+
+  const getCurrentIndices = useCallback(() => {
+    const currentRunIndex = runs.findIndex((run) => run.dag_run_id === runId);
+    const currentTaskIndex = flatNodes.findIndex((node) => node.id === taskId);
+
+    return {
+      runIndex: Math.max(0, currentRunIndex),
+      taskIndex: Math.max(0, currentTaskIndex),
+    };
+  }, [runs, flatNodes, runId, taskId]);
+
+  const navigateToPosition = useCallback(
+    (runIndex: number, taskIndex: number) => {
+      if (runs.length === 0 || flatNodes.length === 0) {
+        return;
+      }
+
+      const run = runs[runIndex];
+      const task = flatNodes[taskIndex];
+
+      if (!run || !task) {
+        return;
+      }
+
+      const { search } = globalThis.location;
+      const searchParams = new URLSearchParams(search);
+      const groupPath = task.isGroup ? "group/" : "";
+      const mappedPath = task.is_mapped ? "/mapped" : "";
+      const path = `/dags/${dagId}/runs/${run.dag_run_id}/tasks/${groupPath}${task.id}${mappedPath}`;
```

Comment:
This logic seems similar to what we have in `TaskLink`, maybe we can extract somme utility function and reuse it at both places to avoid inconsistencies.